internal/image/image.go

```diff
@@ -48,3 +50,38 @@ func ParseWithError(_image string) (string, error) {
 	// https://pkg.go.dev/github.com/docker/distribution/reference?tab=doc#TagNameOnly
 	return reference.TagNameOnly(_canonical).String(), nil
 }
+
+// IsPrivledgedImage digests the provided image with a
+// privledged pattern to see if the image meets the criteria
+// needed to allow a Docker Socket mount.
+func IsPrivledgedImage(image, privledged string) (bool, error) {
+	// parse the image provided into a
+	// named, fully qualified reference
+	//
+	// https://pkg.go.dev/github.com/docker/distribution/reference?tab=doc#ParseAnyReference
+	_refImg, err := reference.ParseNormalizedNamed(image)
+	if err != nil {
+		return false, err
+	}
+
+	// add default tag "latest" when tag does not exist
+	_refImg = reference.TagNameOnly(_refImg)
```

Comment:
I think you could simplify this function by using the functions we already created for parsing an image 👍

https://pkg.go.dev/github.com/go-vela/pkg-runtime/internal/image#ParseWithError

```suggestion
// privledged pattern to see if the image meets the criteria
// needed to allow a Docker Socket mount.
func IsPrivledgedImage(_image, privledged string) (bool, error) {
	// parse the image provided into a named, fully qualified canonical reference
	//
	// https://pkg.go.dev/github.com/go-vela/pkg-runtime/internal/image#ParseWithError
	_refImg, err := image.ParseWithError(_image)
	if err != nil {
		return false, err
	}
```

Reply:
Yeah, I'm not sure I can without making it more complex or changing the return of that function.

In order to use [`FamiliarMatch`](https://pkg.go.dev/github.com/docker/distribution/reference#FamiliarMatch) I need a variable of type [`reference.Reference`](https://github.com/docker/distribution/blob/v2.7.1/reference/reference.go#L65) for the image. 

So, I could modify `ParseWithError` to return me something of that type but it felt easier to just re-use the Docker functions to parse the image.

Reply:
Ahhh I see 👍 

I'm good with creating a custom function to accomplish this (`ParseRawWithError`?) and then use it in the `ParseWithError()` function?

Rough example:

```go
func ParseRawWithError(_image string) (reference.Reference, error) {
	// parse the image provided into a
	// named, fully qualified reference
	//
	// https://pkg.go.dev/github.com/docker/distribution/reference?tab=doc#ParseAnyReference
	_reference, err := reference.ParseAnyReference(_image)
	if err != nil {
		return _image, err
	}

	return _reference, nil
}
```

Otherwise, my only suggestion would be to update the original code for `reference.ParseNormalizedNamed()` to instead call `reference.ParseAnyReference()`.

```go
// privledged pattern to see if the image meets the criteria
// needed to allow a Docker Socket mount.
func IsPrivledgedImage(image, privledged string) (bool, error) {
	// parse the image provided into a
	// named, fully qualified reference
	//
	// https://pkg.go.dev/github.com/docker/distribution/reference?tab=doc#ParseAnyReference
	_refImg, err := reference.ParseAnyReference(image)
	if err != nil {
		return false, err
	}

	// add default tag "latest" when tag does not exist
	_refImg = reference.TagNameOnly(_refImg)
```

Reply:
I think I'd prefer to either add it to the existing function as a third return or do the second option you provided. So, adding it to existing would be like: 
```go
func ParseWithError(_image string) (string, string, error)
```

Then usage kinda reminds me of our [sdk](https://github.com/go-vela/sdk-go/blob/master/vela/build.go#L18) and calling the function would be like:

```go
image, raw, err := internal.ParseWithError(alpine:latest)`
```

Or with what you provided I would also need to call [`ParsedName`](https://pkg.go.dev/github.com/docker/distribution/reference#ParseNamed) to ensure I have the right data type to call -[`TagNameOnly`](https://pkg.go.dev/github.com/docker/distribution/reference#TagNameOnly)

Reply:
What would be the value in adding it as a third return value vs creating a function specifically to handle this?

Typically in Go, messing with the function signature isn't (and shouldn't) be the first resort to handling a problem.

Especially for a function that is prominent like this one where it is used in quite a few spots throughout the `go-vela/pkg-runtime` codebase and that number will likely continue to grow 👍

Reply:
I agree with you changing the signature is not my first choice but I'm having a hard time seeing the value of a creating new function and making it public on the package when it's only used in the one location internal to the package.

Reply:
Hmm... I think the value argument could be made both ways.

Anyways, moving on, I think that leaves us with changing the call to `reference.ParseAnyReference()` then?

Reply:
That works. I'll get this updated then 😄